### PR TITLE
use metal::less as default ordering for metal::sort

### PR DESCRIPTION
**Pull request by ecrypa**
see https://github.com/brunocodutra/metal/issues/73#issuecomment-534259653

### CODE REVIEW

File: include/metal/list/sort.hpp [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
+#include "../lambda/lambda.hpp"
 #include "../list/list.hpp"
 #include "../number/if.hpp"
+#include "../number/less.hpp"
 
 namespace metal {
     /// \cond
     namespace detail {
-        template<typename lbd>
+        template<typename lbd = metal::lambda<metal::less>>
```
Comment (brunocodutra) [inline include/metal/list/sort.hpp]: Looks great, do you think you can add a couple test cases to make sure it can be instantiated with a single argument as long as it's a list of numbers?

I know all taht preprocessor [voodoo](https://github.com/brunocodutra/metal/blob/master/test/unit/include/test/preprocessor.hpp#L133) is inscrutable and could possibly, but I'm happy to help with any questions you may have.

If you're interested, you may want to have a look at [this summary of some of the techniques it uses](https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms) before you dive in.

Comment (brunocodutra) [reply include/metal/list/sort.hpp]: Also, could you add a sentence about this feature to the docs?

File: test/unit/src/metal/list/sort.cpp [hunk 2]
```diff
@@ -3,47 +3,64 @@
...
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), LAMBDA(_)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M)>), (BOOL(M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), VALUE(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), NUMBER(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), PAIR(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LIST(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), MAP(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LAMBDA(N)>), (BOOL(N == 2 || M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LAMBDA(_)>), (BOOL(M < 2))); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M)>), (BOOL(M < 2))); \
```
Comment (ecrypa) [inline test/unit/src/metal/list/sort.cpp]: Caring about the tests made me think about the consequences of this change more thoroughly. In particular, this line for `M = 1` is nasty: It is probably not what the user intends, but it compiles.

I tried to apply the _easy-to-use,-hard-to-misuse_ rule here, but I do not see how we can achieve both (convenient default for the ordering relation, as expected from `std::sort` on the one hand, and exclusion of this nasty case on the other hand). For a moment I considered a solution that treats single-value lists specially (`invoke<lbd, x, x>` just to be sure), but that turns out to be a bad idea.

Do you share my concerns regarding the nasty case? What do you propose?

File: include/metal/list/sort.hpp [hunk 3]
```diff
@@ -29,6 +31,8 @@ namespace metal {
     ///     using result = metal::sort<l, lbd>;
     /// \endcode
     ///
+    /// \tip{`lbd` may be omitted, in which case it defaults to `metal::lambda<metal::less>`.}
```
Comment (ecrypa) [inline include/metal/list/sort.hpp]: This change introduces the first _note_. Here is what `grep -r '\\tip' include/` returns after the change:
```
include/metal/value/value.hpp:    /// \tip{Use `metal::is_value` to trigger \SFINAE.}
include/metal/value/value.hpp:    /// \tip{Use `metal::value` to prevent undesired template pattern matching.}
include/metal/list/sort.hpp:    /// \tip{`lbd` may be omitted, in which case it defaults to `metal::lambda<metal::less>`.}
include/metal/number/if.hpp:    /// \tip{If `num{} != false`, `y` may be omitted.}
include/metal/lambda/bind.hpp:    /// \tip{Use metal::arg<n> as a placeholder for the n-th argument.}
```
The `stable`-sorting information IMHO belongs to another category than these _tips_. On the other hand, the default-argument information seems like a good fit, do you agree?

File: include/metal/list/sort.hpp [hunk 4]
```diff
@@ -46,9 +50,9 @@ namespace metal {
     ///
     /// ### See Also
     /// \see list, reverse, rotate
-    template<typename seq, typename lbd>
+    template<typename seq, typename... lbd>
     using sort = detail::call<
```
Comment (ecrypa) [inline include/metal/list/sort.hpp]: I considered to customize the code shown by doxygen because we wanted it to read
```
template<typename seq, typename lbd = metal::lambda<metal::less>>>
```
and the only reason not to do so is portability for nonconforming compilers. It would surely improve the clarity of the documentation, but let us first decide if we still want the default ordering at all (see my concern in other comment) before we care about this tweak.

File: test/unit/src/metal/list/sort.cpp (hunk 2 above)
Comment (brunocodutra) [reply test/unit/src/metal/list/sort.cpp]: > In particular, this line for `M = 1` is nasty

Very well observed, but this is not an issue that manifests itself only because we added default ordering to `metal::sort` - for an arbitrary `lbd` it is [still true](https://github.com/brunocodutra/metal/blob/master/test/unit/src/metal/list/sort.cpp#L31) that `is_invocable<lambda<sort>, map<pair<void, void>>, lambda<lbd>>::value == true`, even for a nasty `lbd` such as:

```cpp
template<typename... args>
using lbd = static_assert(sizeof...(args)  + 1 == 0, "always fails");
```

If we draw a parallel to runtime algorithms, specifying a function that unconditionally throws an exception as the predicate doesn't mean `std::sort` fails to sort a sequence of size 1.

Based on all that, I'd say this is fine.

File: include/metal/list/sort.hpp (hunk 3 above)
Comment (brunocodutra) [reply include/metal/list/sort.hpp]: Makes sense to me.

File: include/metal/list/sort.hpp (hunk 4 above)
Comment (brunocodutra) [reply include/metal/list/sort.hpp]: Are you sure `template<class seq, class lbd = metal::lambda<metal::less>>>` doesn't work? 
Have you tested? Sometimes MSVC cooperates.

Worst case I agree that we need to customize it for doxygen such that it reads nicely on docs.

File: test/unit/src/metal/list/sort.cpp [hunk 5]
```diff
@@ -3,47 +3,64 @@
...
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), LAMBDA(N)>), (BOOL(N == 2 || M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), LAMBDA(_)>), (BOOL(M < 2))); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort> COMMA(M) NUMBERS(M)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>>), (TRUE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>, VALUE(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>, NUMBER(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>, PAIR(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>, LIST(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>, MAP(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, metal::list<NUMBERS(M)>, metal::list<NUMBERS(N)>>), (FALSE)); \
```
Comment (brunocodutra) [inline test/unit/src/metal/list/sort.cpp]: I'd consider this case covered by `LIST(N)`, but I'm fine with keeping it too.

File: test/unit/src/metal/list/sort.cpp [hunk 6]
```diff
@@ -3,47 +3,64 @@
...
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LAMBDA(_)>), (BOOL(M < 2))); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M)>), (BOOL(M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), VALUE(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), NUMBER(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), PAIR(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), LIST(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), MAP(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), LAMBDA(N)>), (BOOL(N == 2 || M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M), LAMBDA(_)>), (BOOL(M < 2))); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort> COMMA(M) NUMBERS(M)>), (FALSE)); \
```
Comment (brunocodutra) [inline test/unit/src/metal/list/sort.cpp]: I'd omit this one, since I consider it more a concern of `is_invocable` tests to make sure it works for mismatched arity.

File: test/unit/src/metal/list/sort.cpp (hunk 5 above)
Comment (ecrypa) [reply test/unit/src/metal/list/sort.cpp]: A list of numbers is special because the second argument can be omitted. I added all the obvious combinations for symmetry reasons. If there is only the first check (line 46) then it looks incomplete.

Comment (brunocodutra) [reply test/unit/src/metal/list/sort.cpp]: I was referring to line 50, which seems like a more general test case.

File: include/metal/list/sort.hpp (hunk 4 above)
Comment (ecrypa) [reply include/metal/list/sort.hpp]: I had not tested it initially, but in fact there are [failing tests](https://ci.appveyor.com/project/brunocodutra/metal/builds/28065885/job/n97t30q3e0pyaal3#L194) in #107.

Customization of the docs is dangerous though: I can think of a contrived case where our workaround behaves differently. In my mind (and also my search history) there is this error message:
> pack expansion used as argument for non-pack parameter of alias template

What can we do now? I can think of these three options:
 1. try to find another workaround that makes MSVC happy
 1. keep this version and customize the docs (adding a note that the real behavior differs in edge cases)
 1. keep this version, but do not customize the docs

If we choose (3) then the docs are more concise (and contain less potentially confusing notes on contrived edge cases). On the other hand, the clarity of the default argument in code is a strong argument for (2).

What do you think?

File: test/unit/src/metal/list/sort.cpp (hunk 5 above)
Comment (ecrypa) [reply test/unit/src/metal/list/sort.cpp]: I added the test case of line 50 for "symmetry reasons": the test cases are grouped in blocks, and I added a new block. The first argument of the new block is a list of numbers. I forgot to add the list of numbers as a second argument to all other blocks, though.

Anyway, you wrote that you are fine with keeping it. Probably I should learn to read-and-resolve feedback that does not require any actions.

File: include/metal/list/sort.hpp [hunk 7]
```diff
@@ -3,13 +3,15 @@
...
+#include "../lambda/lambda.hpp"
 #include "../list/list.hpp"
 #include "../number/if.hpp"
+#include "../number/less.hpp"
 
 namespace metal {
     /// \cond
     namespace detail {
-        template<class lbd>
+        template<class lbd = metal::lambda<metal::less>>
```
Comment (brunocodutra) [inline include/metal/list/sort.hpp]: This default argument is not necessary, since by the time this is instantiated all arguments are known.
Removing it may help MSVC.

File: include/metal/list/sort.hpp [hunk 8]
```diff
@@ -46,9 +50,9 @@ namespace metal {
     ///
     /// ### See Also
     /// \see list, reverse, rotate
-    template<class seq, class lbd>
+    template<class seq, class... lbd>
     using sort = detail::call<
-        detail::_sort<lbd>::template type,
+        detail::_sort<lbd...>::template type,
```
Comment (brunocodutra) [inline include/metal/list/sort.hpp]: We [do already have default template arguments on aliases](https://github.com/brunocodutra/metal/blob/master/include/metal/value/value.hpp#L91), so there's hope.

File: include/metal/list/sort.hpp (hunk 7 above)
Comment (ecrypa) [reply include/metal/list/sort.hpp]: I am not sure to understand your comment. MSVC emits an ICE for the code in #107. There, the only defaulted parameter is the `lbd` of the alias template.

Comment (brunocodutra) [reply include/metal/list/sort.hpp]: Never mind, my bad, I didn't realize it was already defined this way on that PR.

File: include/metal/list/sort.hpp (hunk 8 above)
Comment (ecrypa) [reply include/metal/list/sort.hpp]: Do you mind if I open a couple of PRs in order to test different versions? Or should I prefer godbolt.org in order to reduce the load of your CI setup?

Comment (brunocodutra) [reply include/metal/list/sort.hpp]: I don't mind extra PRs.

Comment (ecrypa) [reply include/metal/list/sort.hpp]: I tried, but I think it is not worth any more effort. I vote for the [workaround](https://github.com/brunocodutra/metal/pull/101#pullrequestreview-301046192).
